File: mcp/src/transport/wifi.ts
/**
 * WiFi transport: sends PushMessage over raw TCP to a locally discovered iOS device.
 * Matches the Go CLI protocol: JSON bytes written directly to TCP socket, no framing.
 */

import { createConnection, type Socket } from "node:net";
import type { PushMessage, AckMessage } from "../protocol/messages.js";

const CONNECT_TIMEOUT = 10_000;
const WRITE_TIMEOUT = 10_000;
const ACK_READ_TIMEOUT = 10_000;

export interface DeviceInfo {
  name: string;
  host: string;
  port: number;
  id?: string;
}

/** Discover MarkPush devices on local network via mDNS. */
export async function discoverDevice(timeoutMs: number = 2_000): Promise<DeviceInfo | null> {
  // Dynamic import to handle missing native module gracefully.
  const mdns = await import("multicast-dns");
  const instance = mdns.default();

  return new Promise<DeviceInfo | null>((resolve) => {
    const timer = setTimeout(() => {
      instance.destroy();
      resolve(null);
    }, timeoutMs);

    instance.on("response", (response: any) => {
      for (const answer of response.answers) {
        if (answer.type === "PTR" && answer.data?.includes("_markpush._tcp")) {
          // Found a MarkPush service. Look for SRV and A records.
          const srv = response.additionals?.find((a: any) => a.type === "SRV");
          const a = response.additionals?.find((a: any) => a.type === "A");
          const txt = response.additionals?.find((a: any) => a.type === "TXT");

          if (srv && a) {
            clearTimeout(timer);
            instance.destroy();

            let deviceId: string | undefined;
            if (txt?.data) {
              const txtStr = Buffer.isBuffer(txt.data[0])
                ? txt.data.map((b: Buffer) => b.toString()).join("")
                : String(txt.data);
              const idMatch = txtStr.match(/id=([^\s]+)/);
              if (idMatch) deviceId = idMatch[1];
            }

            resolve({
              name: srv.data?.target || "MarkPush Device",
              host: a.data,
              port: srv.data?.port || 49152,
              id: deviceId,
            });
            return;
          }
        }
      }
    });

    // Query for MarkPush services.
    instance.query({
      questions: [{ name: "_markpush._tcp.local", type: "PTR" }],
    });
  });
}

/** Send a PushMessage over raw TCP to a device (matches Go CLI protocol). */
export async function sendViaWiFi(device: DeviceInfo, msg: PushMessage): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const socket: Socket = createConnection(
      { host: device.host, port: device.port, timeout: CONNECT_TIMEOUT },
      () => {
        // Connected — write JSON payload directly (no framing, matching Go CLI).
        const payload = JSON.stringify(msg);
        socket.write(payload, "utf-8", (err) => {
          if (err) {
            socket.destroy();
            reject(new Error(`WiFi send failed: ${err.message}`));
            return;
          }
        });
      },
    );

    // Read ACK (best-effort, matching Go CLI behavior).
    const chunks: Buffer[] = [];
    const ackTimer = setTimeout(() => {
      socket.destroy();
      resolve(); // No ACK within timeout — message was still sent.
    }, ACK_READ_TIMEOUT);

    socket.on("data", (data) => {
      chunks.push(data);
      clearTimeout(ackTimer);

      try {
        const ack: AckMessage = JSON.parse(Buffer.concat(chunks).toString("utf-8"));
        if (ack.status === "error") {
          socket.destroy();
          reject(new Error(`Device reported error for message ${msg.id}`));
          return;
        }
      } catch {
        // Partial data or parse failure — non-fatal.
      }

      socket.destroy();
      resolve();
    });

    socket.on("timeout", () => {
      socket.destroy();
      reject(new Error(`WiFi connection timed out to ${device.host}:${device.port}`));
    });

    socket.on("error", (err) => {
      clearTimeout(ackTimer);
      reject(new Error(`WiFi connection error: ${err.message}`));
    });

    socket.on("close", () => {
      clearTimeout(ackTimer);
      resolve(); // If closed without ACK, message was still sent.
    });
  });
}

File: cli/internal/transport/wifi.go
package transport

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"time"

	"github.com/rahilsinghi/markpush/cli/internal/mdns"
	"github.com/rahilsinghi/markpush/cli/internal/protocol"
)

const (
	// DefaultDiscoveryTimeout is the time to wait for mDNS discovery.
	DefaultDiscoveryTimeout = 2 * time.Second
	// WriteTimeout is the TCP write deadline.
	WriteTimeout = 10 * time.Second
)

// WiFiSender sends messages over local WiFi via raw TCP.
// It discovers the iOS app on the network using mDNS, then connects
// via TCP to push the message as JSON.
type WiFiSender struct {
	DiscoveryTimeout time.Duration
	// device can be pre-set to skip discovery (used by auto-select).
	device *mdns.Device
}

// NewWiFiSender creates a WiFi transport with default settings.
func NewWiFiSender() *WiFiSender {
	return &WiFiSender{
		DiscoveryTimeout: DefaultDiscoveryTimeout,
	}
}

// NewWiFiSenderWithDevice creates a WiFi transport targeting a specific device.
func NewWiFiSenderWithDevice(dev *mdns.Device) *WiFiSender {
	return &WiFiSender{
		DiscoveryTimeout: DefaultDiscoveryTimeout,
		device:           dev,
	}
}

// Send discovers the iOS device (if not already known), connects via TCP,
// sends the push message as JSON, and waits for an acknowledgment.
func (t *WiFiSender) Send(ctx context.Context, msg *protocol.PushMessage) error {
	dev := t.device
	if dev == nil {
		var err error
		dev, err = mdns.Discover(ctx, t.DiscoveryTimeout)
		if err != nil {
			return fmt.Errorf("wifi send: %w", err)
		}
	}

	addr := fmt.Sprintf("%s:%d", dev.Host, dev.Port)
	dialer := net.Dialer{Timeout: WriteTimeout}
	conn, err := dialer.DialContext(ctx, "tcp", addr)
	if err != nil {
		return fmt.Errorf("wifi send: connect to %s: %w", addr, err)
	}
	defer conn.Close()

	if err := conn.SetWriteDeadline(time.Now().Add(WriteTimeout)); err != nil {
		return fmt.Errorf("wifi send: set deadline: %w", err)
	}

	payload, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("wifi send: marshal message: %w", err)
	}

	if _, err := conn.Write(payload); err != nil {
		return fmt.Errorf("wifi send: write message: %w", err)
	}

	// Wait for acknowledgment.
	if err := conn.SetReadDeadline(time.Now().Add(WriteTimeout)); err != nil {
		return fmt.Errorf("wifi send: set read deadline: %w", err)
	}

	buf := make([]byte, 4096)
	n, err := conn.Read(buf)
	if err != nil {
		// Not fatal — message was sent, ack is best-effort.
		return nil
	}

	var ack protocol.AckMessage
	if err := json.Unmarshal(buf[:n], &ack); err != nil {
		return nil // ack parsing failure is non-fatal
	}

	if ack.Status == "error" {
		return fmt.Errorf("wifi send: device reported error for message %s", msg.ID)
	}

	return nil
}
